File: apps/api/schemas/transfer.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from schemas.squad import SquadResponse


# ---------------------------------------------------------------------------
# Requests
# ---------------------------------------------------------------------------

class TransferPairRequest(BaseModel):
    """One player out + one player in."""
    player_out_id: int = Field(description="ID of the player to remove from your squad.")
    player_in_id: int = Field(description="ID of the replacement player to add.")


class TransferPreviewRequest(BaseModel):
    """Preview 1–5 transfer pairs without applying them."""
    transfers: list[TransferPairRequest] = Field(
        min_length=1,
        max_length=5,
        description="Up to 5 transfer pairs to preview as a batch.",
    )


class TransferApplyRequest(BaseModel):
    """Apply 1–5 transfer pairs atomically."""
    transfers: list[TransferPairRequest] = Field(
        min_length=1,
        max_length=5,
        description="Up to 5 transfer pairs to apply.",
    )


# ---------------------------------------------------------------------------
# Shared sub-objects
# ---------------------------------------------------------------------------

class PlayerBriefResponse(BaseModel):
    id: int
    name: str
    display_name: Optional[str]
    position: str
    team_name: str
    team_short_name: str
    current_price: float

    model_config = {"from_attributes": False}


class TransferPairPreview(BaseModel):
    """One transfer pair with prices and cost information."""
    player_out: PlayerBriefResponse
    player_in: PlayerBriefResponse
    price_out: float
    price_in: float
    budget_delta: float     # price_out - price_in (positive = profit, negative = cost)
    is_free: bool
    point_cost: int

    model_config = {"from_attributes": False}


# ---------------------------------------------------------------------------
# Responses
# ---------------------------------------------------------------------------

class TransferPreviewResponse(BaseModel):
    """
    What would happen if these transfers were applied.

    If errors is non-empty, nothing should be applied.
    is_valid: convenience flag — True when errors is empty.
    """
    gameweek_id: int
    gameweek_number: int
    is_editable: bool
    transfers: list[TransferPairPreview]
    budget_before: float
    budget_after: float
    total_points_hit: int
    free_transfers_before: int
    free_transfers_after: int
    is_valid: bool
    errors: list[str]

    model_config = {"from_attributes": False}


class TransferApplyResponse(BaseModel):
    """
    Confirmation after transfers are applied.

    Returns the updated squad alongside the applied transfer details
    so the frontend can update all views in one response.
    """
    applied_transfers: list[TransferPairPreview]
    squad: SquadResponse
    total_points_hit: int
    free_transfers_remaining: int
    transfers_this_gw: int

    model_config = {"from_attributes": False}


class TransferSummaryResponse(BaseModel):
    """
    Current gameweek transfer status for the authenticated user.
    Returned by GET /transfers/summary.
    """
    gameweek_id: int
    gameweek_number: int
    gameweek_name: str
    gameweek_deadline: datetime
    gameweek_status: str
    is_editable: bool
    free_transfers_available: int
    transfers_made_this_gw: int
    points_hit_this_gw: int
    budget_remaining: float

    model_config = {"from_attributes": False}


class TransferHistoryItemResponse(BaseModel):
    """One past transfer record."""
    id: int
    gameweek_number: int
    player_out_id: int
    player_out_name: str
    player_in_id: int
    player_in_name: str
    price_out: float
    price_in: float
    is_free: bool
    point_cost: int
    created_at: datetime

    model_config = {"from_attributes": False}


class TransferHistoryResponse(BaseModel):
    transfers: list[TransferHistoryItemResponse]
    total: int
    page: int
    per_page: int
    total_pages: int

    model_config = {"from_attributes": False}

File: apps/web/src/lib/api/transfer.ts
import { apiFetch, ApiError } from './client'
import type { SquadResponse } from './squad'

export type { ApiError }

// ---------------------------------------------------------------------------
// Shared types
// ---------------------------------------------------------------------------

export interface PlayerBrief {
  id: number
  name: string
  display_name: string | null
  position: string
  team_name: string
  team_short_name: string
  current_price: number
}

export interface TransferPairPreview {
  player_out: PlayerBrief
  player_in: PlayerBrief
  price_out: number
  price_in: number
  budget_delta: number
  is_free: boolean
  point_cost: number
}

// ---------------------------------------------------------------------------
// Summary
// ---------------------------------------------------------------------------

export interface TransferSummaryResponse {
  gameweek_id: number
  gameweek_number: number
  gameweek_name: string
  gameweek_deadline: string
  gameweek_status: string
  is_editable: boolean
  free_transfers_available: number
  transfers_made_this_gw: number
  points_hit_this_gw: number
  budget_remaining: number
}

export async function fetchTransferSummary(): Promise<TransferSummaryResponse | null> {
  try {
    return await apiFetch<TransferSummaryResponse>('/api/v1/transfers/summary')
  } catch (err) {
    if (err instanceof ApiError && (err.status === 404 || err.status === 503)) return null
    throw err
  }
}

// ---------------------------------------------------------------------------
// Preview
// ---------------------------------------------------------------------------

export interface TransferPreviewResponse {
  gameweek_id: number
  gameweek_number: number
  is_editable: boolean
  transfers: TransferPairPreview[]
  budget_before: number
  budget_after: number
  total_points_hit: number
  free_transfers_before: number
  free_transfers_after: number
  is_valid: boolean
  errors: string[]
}

export interface TransferPairRequest {
  player_out_id: number
  player_in_id: number
}

export async function previewTransfers(
  pairs: TransferPairRequest[],
): Promise<TransferPreviewResponse> {
  return apiFetch<TransferPreviewResponse>('/api/v1/transfers/preview', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ transfers: pairs }),
  })
}

// ---------------------------------------------------------------------------
// Apply
// ---------------------------------------------------------------------------

export interface TransferApplyResponse {
  applied_transfers: TransferPairPreview[]
  squad: SquadResponse
  total_points_hit: number
  free_transfers_remaining: number
  transfers_this_gw: number
}

export async function applyTransfers(
  pairs: TransferPairRequest[],
): Promise<TransferApplyResponse> {
  return apiFetch<TransferApplyResponse>('/api/v1/transfers/apply', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ transfers: pairs }),
  })
}

// ---------------------------------------------------------------------------
// History
// ---------------------------------------------------------------------------

export interface TransferHistoryItem {
  id: number
  gameweek_number: number
  player_out_id: number
  player_out_name: string
  player_in_id: number
  player_in_name: string
  price_out: number
  price_in: number
  is_free: boolean
  point_cost: number
  created_at: string
}

export interface TransferHistoryResponse {
  transfers: TransferHistoryItem[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

export async function fetchTransferHistory(
  page = 1,
  perPage = 20,
): Promise<TransferHistoryResponse> {
  return apiFetch<TransferHistoryResponse>(
    `/api/v1/transfers/history?page=${page}&per_page=${perPage}`,
  )
}
